### DISCUSSION TITLE
Gradient of 2D Mesh

### DISCUSSION THREAD

**Discussion by 19kdc3** [marcomusy/vedo#634] [Q&A] [answered]
Hi Marco,

I have a 2D mesh that exists in the XZ-plane and would like to find the gradient within/on the mesh, towards the outside/edge. The `.operation("gradient")` seems to only calculate normal to the mesh surface.

![image](https://user-images.githubusercontent.com/73371848/165394960-1d914ae7-cbec-4b15-b5da-7a5830a00038.png)

![image](https://user-images.githubusercontent.com/73371848/165396657-185a6c33-d714-4d77-b8c7-03c3346fa251.png)

Alternatively, I've tried using `.silhouette('2d')` since it mentions returning a "new line mesh", but I can't seem to do any thing more once a silhouette is applied to the mesh.

![image](https://user-images.githubusercontent.com/73371848/165396378-2fdb4427-142e-43dc-af3e-785d17d2b86e.png)

 Is there a way to use the gradient operation for this use case?

Thanks for the help!

**Comment by marcomusy** [comment, marked as answer]
hi, not sure if i understand the question (gradient of what scalar field? the distance from the boundary?), if so you can maybe compute it with `mesh.distanceTo(silhuette)` where `silhuette=mesh.silhouette('2d')`

**Reply by 19kdc3** [reply]
Yes, I'm looking for the gradient of the distance from the 2D mesh to the silhouette. Similar to what you did here, but in 2D:
[https://github.com/marcomusy/vedo/issues/632#issuecomment-1107448700](url)
 
I tried implementing your suggestion "`mesh.distanceTo(silhuette)` where `silhuette=mesh.silhouette('2d')`", but I simply get this:
![image](https://user-images.githubusercontent.com/73371848/165402191-1137c9ab-1f0f-442d-8293-6bc4047dde50.png)

```
    sil = mesh.silhouette('2d').c('r')
    mesh.distanceTo(sil)
    mesh.cmap('hot').addScalarBar('Signed\nDistance')

    vd.show(mesh, sil, N=2, axes=1, viewup='z').close()
```

**Reply by marcomusy** [reply]
use boundaries instead, eg:
```python
from vedo import *

shape = load(dataurl+'timecourse1d.npy')[56]

msh = shape.tomesh(invert=1).smooth()             

bnd = msh.boundaries().c('g5').lw(5)
msh.distanceTo(bnd)
msh.cmap('hot').addScalarBar()

show(msh, bnd, bg='bb').close()
```
<img width="505" alt="Screenshot 2022-04-27 at 00 41 39" src="https://user-images.githubusercontent.com/32848391/165404554-6ec9a282-55da-4bd6-b078-1756bf930a6b.png">

**Reply by marcomusy** [reply]
..for the gradient:
```python
from vedo import *

shape = load(dataurl+'timecourse1d.npy')[56]

msh = shape.tomesh(invert=1).smooth().lw(0)           

bnd = msh.boundaries().c('g5').lw(5)
msh.distanceTo(bnd)
msh.cmap('hot').addScalarBar()

gradd = msh.gradient()
arrs = Arrows2D(msh.points(), msh.points()-gradd*0.3)
arrs.z(0.01) # so the stay on top
arrs.c('white')

show(msh, bnd, arrs, bg='bb').close()
```

<img width="371" alt="Screenshot 2022-04-27 at 00 58 51" src="https://user-images.githubusercontent.com/32848391/165406324-547960f1-67fd-443c-9893-ba575ff1db0b.png">

**Reply by 19kdc3** [reply]
That looks great! From those pictures I believe that's what I'm looking for. 

Unfortunately, when I try to run your code I get the error: "'Mesh' object has no attribute 'GetScalars'" for the line `gradd = msh.gradient()`

Do you have any suggestions?

**Reply by marcomusy** [reply]
oh, have you tried
`pip install vedo -U`

ps : careful about `mesh` and `msh` because you are importing with `from vedo import *`

**Reply by 19kdc3** [reply]
I thought I had the most recent version, since I installed it recently, but I guess not recent enough! 

Thanks so much for the help! I was able to achieve the desired results:
![image](https://user-images.githubusercontent.com/73371848/165413433-24f690c1-b8ec-4898-af4f-4622c55a1296.png)
![image](https://user-images.githubusercontent.com/73371848/165413450-e498c634-cd91-4933-87fa-e56a2358bc7b.png)
![image](https://user-images.githubusercontent.com/73371848/165413461-d16f2d8b-6a70-4932-bfb2-64f2ddea1f3b.png)

A small inconvenience is that the arrows are drawn on the wrong plane such that they appear to have "zero thickness" when viewed in the y-axis. Is there an easy way to correct that?

If I'd like to probe for values throughout the gradient (not necessarily at a node/vert location) does the method you provided allow for interpolation? or does it need to switch from a mesh to volume?

Again, thanks so much for the help! I really appreciate it.

**Reply by marcomusy** [reply]
Nice! 
About arrows, just use the 3d version `Arrows()` instead.
About interpolation check out https://github.com/marcomusy/vedo/blob/master/examples/advanced/interpolateScalar4.py and the other of the serie.
